From fc70d537fc5337806426a9cb0e88f97b75344aba Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Sun, 24 Jan 2010 16:28:55 +0000
Subject: [PATCH] [CONFIGURATION-405] Make sure that the root node of
 XMLPropertyListConfiguration is a PListNode. Otherwise a
 ConfigurationException is thrown when loading configuration files that do not
 have an outer dict element.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@902596 13f79535-47bb-0310-9956-ffa450edef68
---
 conf/test2.plist.xml                          | 13 ++++++
 .../plist/XMLPropertyListConfiguration.java   | 17 ++++++++
 .../TestXMLPropertyListConfiguration.java     | 42 ++++++++++++++++---
 xdocs/changes.xml                             |  4 ++
 4 files changed, 70 insertions(+), 6 deletions(-)
 create mode 100644 conf/test2.plist.xml

diff --git a/conf/test2.plist.xml b/conf/test2.plist.xml
new file mode 100644
index 0000000000..03a25574a6
--- /dev/null
+++ b/conf/test2.plist.xml
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist SYSTEM "file://localhost/System/Library/DTDs/PropertyList.dtd">
+<plist version="1.0">
+  <array>
+    <array>
+      <string>http://www.google.com/search?hl=en&amp;client=safari&amp;rls=en&amp;q=RFC822MessageDatasPboardType&amp;btnG=Search&amp;aq=f&amp;oq=&amp;aqi=</string>
+    </array>
+    <array>
+      <string>RFC822MessageDatasPboardType - Google Search</string>
+    </array>
+  </array>
+</plist>
+
diff --git a/src/java/org/apache/commons/configuration/plist/XMLPropertyListConfiguration.java b/src/java/org/apache/commons/configuration/plist/XMLPropertyListConfiguration.java
index 7d75b295e2..190caf49c9 100644
--- a/src/java/org/apache/commons/configuration/plist/XMLPropertyListConfiguration.java
+++ b/src/java/org/apache/commons/configuration/plist/XMLPropertyListConfiguration.java
@@ -136,6 +136,7 @@ public class XMLPropertyListConfiguration extends AbstractHierarchicalFileConfig
      */
     public XMLPropertyListConfiguration()
     {
+        initRoot();
     }
 
     /**
@@ -224,6 +225,14 @@ public void addProperty(String key, Object value)
 
     public void load(Reader in) throws ConfigurationException
     {
+        // We have to make sure that the root node is actually a PListNode.
+        // If this object was not created using the standard constructor, the
+        // root node is a plain Node.
+        if (!(getRootNode() instanceof PListNode))
+        {
+            initRoot();
+        }
+
         // set up the DTD validation
         EntityResolver resolver = new EntityResolver()
         {
@@ -414,6 +423,14 @@ else if (value != null)
         }
     }
 
+    /**
+     * Helper method for initializing the configuration's root node.
+     */
+    private void initRoot()
+    {
+        setRootNode(new PListNode());
+    }
+
     /**
      * SAX Handler to build the configuration nodes while the document is being parsed.
      */
diff --git a/src/test/org/apache/commons/configuration/plist/TestXMLPropertyListConfiguration.java b/src/test/org/apache/commons/configuration/plist/TestXMLPropertyListConfiguration.java
index 75df6639f0..a6b3a56084 100644
--- a/src/test/org/apache/commons/configuration/plist/TestXMLPropertyListConfiguration.java
+++ b/src/test/org/apache/commons/configuration/plist/TestXMLPropertyListConfiguration.java
@@ -18,17 +18,22 @@
 package org.apache.commons.configuration.plist;
 
 import java.io.File;
-import java.util.*;
+import java.util.Calendar;
+import java.util.Iterator;
+import java.util.List;
+import java.util.TimeZone;
 
 import junit.framework.TestCase;
-import junitx.framework.ObjectAssert;
 import junitx.framework.ArrayAssert;
 import junitx.framework.ListAssert;
-import org.apache.commons.configuration.FileConfiguration;
+import junitx.framework.ObjectAssert;
+
 import org.apache.commons.configuration.Configuration;
+import org.apache.commons.configuration.ConfigurationComparator;
+import org.apache.commons.configuration.ConfigurationException;
+import org.apache.commons.configuration.FileConfiguration;
 import org.apache.commons.configuration.HierarchicalConfiguration;
 import org.apache.commons.configuration.StrictConfigurationComparator;
-import org.apache.commons.configuration.ConfigurationComparator;
 
 /**
  * @author Emmanuel Bourg
@@ -279,7 +284,7 @@ public void testSaveEmptyDictionary() throws Exception
         {
             assertTrue(savedFile.delete());
         }
-        
+
         // save the configuration
         String filename = savedFile.getAbsolutePath();
         config.save(filename);
@@ -288,7 +293,7 @@ public void testSaveEmptyDictionary() throws Exception
 
         // read the configuration and compare the properties
         Configuration checkConfig = new XMLPropertyListConfiguration(new File(filename));
-        
+
         assertEquals(null, config.getProperty("empty-dictionary"));
         assertEquals(null, checkConfig.getProperty("empty-dictionary"));
     }
@@ -336,4 +341,29 @@ public void testInitCopy()
 		StrictConfigurationComparator comp = new StrictConfigurationComparator();
 		assertTrue("Configurations are not equal", comp.compare(config, copy));
 	}
+
+    /**
+     * Tests whether a configuration can be loaded that does not start with a
+     * <code>dict</code> element. This test case is related to
+     * CONFIGURATION-405.
+     */
+    public void testLoadNoDict() throws ConfigurationException
+    {
+        XMLPropertyListConfiguration plist = new XMLPropertyListConfiguration();
+        plist.setFileName("conf/test2.plist.xml");
+        plist.load();
+        assertFalse("Configuration is empty", plist.isEmpty());
+    }
+
+    /**
+     * Tests whether a configuration that does not start with a
+     * <code>dict</code> element can be loaded from a constructor. This test
+     * case is related to CONFIGURATION-405.
+     */
+    public void testLoadNoDictConstr() throws ConfigurationException
+    {
+        XMLPropertyListConfiguration plist = new XMLPropertyListConfiguration(
+                "conf/test2.plist.xml");
+        assertFalse("Configuration is empty", plist.isEmpty());
+    }
 }
diff --git a/xdocs/changes.xml b/xdocs/changes.xml
index 4cfff37bc6..9523a991e4 100644
--- a/xdocs/changes.xml
+++ b/xdocs/changes.xml
@@ -23,6 +23,10 @@
 
   <body>
     <release version="1.7" date="in SVN" description="">
+      <action dev="oheger" type="fix" issue="CONFIGURATION-405">
+        XMLPropertyListConfiguration no longer throws a ConfigurationException
+        if the file to be loaded does not have an outer dict element.
+      </action>
       <action dev="oheger" type="fix" issue="CONFIGURATION-403">
         When an empty XMLConfiguration was saved and reloaded the root element
         was assigned an empty text value. Because of this isEmpty() returned
